### ISSUE TITLE
A unified way of providing global/thread-local configurations

### ISSUE THREAD

**Issue by beam2d** [pfnet/chainer#2048]
Some behaviors of Chainer could be configured by a global or thread-local configuration (often as a flag). We already have two flags and one integer configuration: default_backprop (which is configured by no_backprop_mode and force_backprop_mode), type_check_enable (which is configured by CHAINER_TYPE_CHECK environment variable), and cuda._max_workspace_size (which is configured by cuda.set_max_workspace_size). We may have more flags and values to configure the behavior of features which are currently configured by arguments to certain functions. Here is a list of features configured by global/local configurations (not exhaustive).

- use_cudnn flag in various functions.
- train/test flag in some links and functions (e.g. dropout and BatchNormalization)
- deterministic flag in some convolution functions

These would be easier to configure by a global or thread-local flag instead of local arguments. These can be implemented in a similar way as done for existing flags, but it is beneficial to provide a unified way of configuring them, because it reduces the amount of things to learn for using Chainer.

Here is a simple design proposal.

- Add a central global object (_configuration_) to manage the configuration. It has a global dictionary that gives the default configuration and a thread-local dictionary that gives the thread-wise configuration.
- A flag is registered to the configuration object with the flag name and the default value. The flag name is automatically converted to the corresponding environment variable name (e.g. enable_foo is converted to CHAINER_ENABLE_FOO) and the environment variable is used as the default value if given.
- The configuration object has interface to get the value for the current thread, set the value for the current thread, and set the value for the global setting.

A context manager function for temporarily changing the configuration would be also useful (like no_backprop_mode function).

**Comment by delta2323** [comment]
Should we consider this issue as resolved?

**Comment by beam2d** [comment]
Not yet. Some concrete examples (use_cudnn and deterministic) are not made or merged.

**Comment by delta2323** [comment]
OK. Now `training` and `use_cudnn` mode are done, the remaining one is `deterministic` mode isn't it?

**Comment by beam2d** [comment]
Yes, the deterministic flag remains.

**Comment by delta2323** [comment]
All configurations discussed so far have been implemented. So I close this PR.
